pkg/libp2p/peermgr/peermgr.go

```diff
@@ -0,0 +1,289 @@
+package peermgr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/libp2p/go-libp2p/core/host"
+	net "github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/rs/zerolog/log"
+	"golang.org/x/exp/slices"
+)
+
+type PeerMgr struct {
+	// the libp2p host
+	h host.Host
+
+	// peers we are instructing peermrg to maintain connections with.
+	bootstrappers []multiaddr.Multiaddr
+
+	// signals service is currently bootstrapping to bootstrap peers.
+	bootstrapping chan struct{}
+
+	// peers we are currently connected to.
+	peersMu sync.Mutex
+	peers   map[peer.ID]struct{}
+
+	notifee *net.NotifyBundle
+
+	// signals service exit
+	done chan struct{}
+
+	cfg config
+}
+
+func New(h host.Host, bootstrap []multiaddr.Multiaddr, opts ...Option) (*PeerMgr, error) {
+	cfg := &config{
+		runInterval:      time.Second * 5,
+		bootstrapTimeout: time.Second * 30,
+		// default is a real clock
+		clock: clock.New(),
+	}
+	for _, opt := range opts {
+		if err := opt(cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	pm := &PeerMgr{
+		h:             h,
+		bootstrappers: bootstrap,
+		peers:         make(map[peer.ID]struct{}),
+		done:          make(chan struct{}),
+		bootstrapping: make(chan struct{}, 1),
+		cfg:           *cfg,
+	}
+
+	// register notification for peer dis/connection, used to maintain peers set.
+	pm.notifee = &net.NotifyBundle{
+		// triggered whenever we disconnect from a peer
+		DisconnectedF: func(_ net.Network, conn net.Conn) {
+			pm.handleDisconnect(conn.RemotePeer())
+		},
+		// triggered whenever we connect to a peer
+		ConnectedF: func(_ net.Network, conn net.Conn) {
+			pm.handleConnect(conn.RemotePeer())
+		},
+	}
+
+	h.Network().Notify(pm.notifee)
+
+	return pm, nil
+}
+
+func (m *PeerMgr) PeerCount() int {
+	m.peersMu.Lock()
+	defer m.peersMu.Unlock()
+	return len(m.peers)
+}
+
+func (m *PeerMgr) Start(ctx context.Context) {
+	log.Ctx(ctx).Info().Msg("starting peermgr")
+	go m.Run(ctx)
+}
+
+func (m *PeerMgr) Stop(ctx context.Context) {
+	log.Ctx(ctx).Info().Msg("stopping peermgr")
+	close(m.done)
+}
+
+func (m *PeerMgr) Run(ctx context.Context) {
+	// eagerly do bootstrapping at startup, after initial run look interval
+	m.doBootstrap(ctx)
+	tick := m.cfg.clock.Ticker(m.cfg.runInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
```

Comment:
> This defaults to 5 seconds.

I should have made this 10 seconds remain in line with previous behavior: https://github.com/bacalhau-project/bacalhau/blob/main/pkg/libp2p/host.go#L24 will address before merging.

>  So we're saying that every 5 seconds we'll try to reconnect to all of the addresses in the bootstrappers list.

Iff we are not connected to all peers in the bootstrap list, yes, otherwise `Run` is a noop. see lines 107-109 below.

> Is this intended to act as a keep-alive for non-TCP connections?

It's intended to ensure we are always connected to bootstrap peers in the event a connection is lost - e.g. node bounces, network outages, IP address behind DNS changes, cats chewing through Ethernet cables in server rooms, etc.

>  I guess a later optimisation is to only reconnect to the ones we know are missing?

Yes, I suppose that could be a follow on to this, although iirc Libp2p doesn't re-dial peers it's already connected to. In the event `Connect` is called on a already connected peer, the operation is a noop. Although, this won't be the case when using non-p2p multiaddresses as we will calll the API to get a pubkey each time regardless of any existing connection. I will add a comment stating this, but generally suspect it won't be much of an issue as far as performance.